### ISSUE TITLE
KinematicBody2D move leaks some unexpected motion colliding with static

### ISSUE THREAD

**Issue by ravenblackx** [godotengine/godot#8603]
**Operating system or device - Godot version:**
Windows, 64-bit, v2.1.3.stable.official

**Issue description:**
Trying to make a kinematic character stick to the ground when walking (so as to not lift off due to momentum when walking fast over bumps).
Working method is to move horizontally, add slide, then move downwards a little faster than the fastest possible walking speed, and if that doesn't collide, undo it and switch to "falling" mode because you've walked off a cliff not down a slope. :)
This latter downward move is *not* followed by a slide, but nevertheless results in the character slowly moving downhill when trying to stand still on a slope. The vectors returned by the various move calls are (0, 0) from the "move horizontally no distance" move, then (0, 3.27ish) from the floor-sticking vertical move - so it's suggested by these return values that no horizontal movement is occurring, but nevertheless the character slowly moves horizontally.

**Steps to reproduce:**
Make a 45 degree diagonal sloping static collider and a kinetic body that acts this way. Try to stand the kinetic body still on the slope. Sample attached.
The cut down code segment:
```
const WALK_SPEED = 200.0
var velocity = Vector2()

func _fixed_process(delta):
		velocity.y = 0
		if (Input.is_action_pressed(global.MOVE_LEFT)):
			velocity.x = - WALK_SPEED
		elif (Input.is_action_pressed(global.MOVE_RIGHT)):
			velocity.x =   WALK_SPEED
		else:
			velocity.x = 0.0
		var motion = velocity * delta
		motion = move(motion)
		if (is_colliding()):
			var n = get_collision_normal()
			motion = n.slide(motion)
			move(motion)
		# stick to the ground
		motion = Vector2(0, WALK_SPEED*1.01*delta)
		motion = move(motion)
		if (!is_colliding()):
			revert_motion()
			print("falling")
			state = STATE.AIRBORNE
		else:
			# fix the leak of x-motion
			print(motion)
```

**Link to minimal example project:**
[Megajoust.zip](https://github.com/godotengine/godot/files/967294/Megajoust.zip)

**Comment by eon-s** [comment]
I don't know if there is a bug here but steep slopes are hard to manage in general (more the downward ones), I usually change the mechanic to work with them, like using the floor normal to affect motion (input+gravity) if I want to stick the body to the ground.

**Comment by ravenblackx** [comment]
Yes, I recognize it's tricky, that's why there's a "stick to the ground" section of code to cover it. I have managed to make a workaround that hides the problem, wherein I don't make the stick to the ground move if I already saw I was going uphill (thus eliminating the problem caused by the janky move while walking uphill), or if I was not moving horizontally anyway (thus eliminating the problem caused by the janky move introducing slow motion on an unrequested axis), but I think it's still a bug.

I'm guessing the collision is being resolved at some point by pushing the object back out in the direction of the normal, rather than by cutting the requested movement short before it causes any intersection.

**Comment by reduz** [comment]
It does both things, it pushes the object away and cuts the movement short.

Just cutting the movement short can easily result in intersections anyway
or numerical precision issues.

Current method ensures things just work and that your body never gets stuck
no matter what (which is a way more complex situation to handle), but you
have to hack some simpler situations manually for them to work. This is why
3.0 introduces move_and_slide, which solves most of then.

For sticking to the floor, I wonder what is best, i thought the easiest
hack is just doing some raycast down and check if there is a floor up to a
given distance.. then stick the character to it

**Comment by eon-s** [comment]
The thing is, (a bad example) when somebody walk down a steep slope, the motion changes, the horizontal value is slowed down or the person will have some little jumps or fall, the same effect needs to be applied to characters, it can't keep the same motion if the horizontal speed is too high.

On smaller slopes the effect could be less noticeable (always depends on the relation gravity/speed).

@reduz I don't know if methods like `move_and_slide` should take in consideration steep slopes or must be left to the mechanic designer if reach that situation.

A method could be restricting the motion (component) normal to the gravity when floor normal is too steep.

**Comment by ravenblackx** [comment]
Maybe a clarification in the documentation of move() would be enough. It really sounds from the description like the mechanism used to avoid intersection is just cutting the requested move short - if I had known up front that it was doing the same kludgy-shove resolution as other physics engines use I wouldn't have spent a couple of hours trying to figure out where *I* was letting some horizontal movement into my script.  (move_and_slide is the opposite of the solution I want!)

I think for my purposes I might be better off using `Physics2DDirectSpaceState.cast_motion` and thereby making sure my `move` calls just never collide.

**Comment by eon-s** [comment]
@ravenblackx maybe `test_move` on sliding to make sure it stays on floor (I ignore the "cost" of constant `test_move`), stop sliding/reduce horizontal movement if not.

**Comment by RandomShaper** [comment]
Not sure if this is related to your issue, but there was a bug that made every motion or motion test assume the same initial position. That is, instead of accumulating the position, the physics backend was always considering the motion initiated on the location the character was at the start of a frame. 

That can give sad surprises if doing multiple motion calls during a frame for finely tuned movement. I was facing issues with that until I found the bug.

Now you have a project setting (Physics 2D -> enable_motion_fix, or something similar) that corrects that behavior. It's off by default to avoid breaking projects already developed relying on the buggy behavior. For 3.0 that lack of motion accumulation has been totally got ridden of.

**Comment by eon-s** [comment]
@RandomShaper I'm sure that bug does not help here. 
I was not aware of that option, I will try a few old KB bugs that could be related to that.

**Comment by kubecz3k** [comment]
First of all thank you for your report and sorry for the delay.

We released Godot 3.0 in January 2018 after 18 months of work, fixing many old issues either directly, or by obsoleting/replacing the features they were referring to.

We still have hundreds of issues whose relevance/reproducibility needs to be checked against the [current stable version](https://godotengine.org/download), and that's where you can help us.
**Could you check if the issue that you described initially is still relevant/reproducible in Godot 3.0 or any newer version, and comment about its current status here?**

For bug reports, please also make sure that the issue contains detailed steps to reproduce the bug and, if possible, a zipped project that can be used to reproduce it right away. This greatly speeds up debugging and bugfixing tasks for our contributors.

Our Bugsquad will review this issue more in-depth in 15 days, and potentially close it if its relevance could not be confirmed.

Thanks in advance.

*Note: This message is being copy-pasted to many "stale" issues (90+ days without activity). It might happen that it is not meaningful for this specific issue or appears oblivious of the issue's context, if so please comment to notify the Bugsquad about it.*

**Comment by akien-mga** [comment]
As there was no answer to the above request, we're closing this issue as inactive.
If you can still reproduce it, please open a new issue against a currently supported release of Godot.
